### PR TITLE
5450 use stock movement status for stock out check

### PR DESCRIPTION
**Pull request by jniles**
This PR refactors the stock out checks to use the `stock_movement_status` table.  This improves the performance by not needing to scan the entire stock_movement table.

Partially addresses #5450.

### CODE REVIEW

File: client/src/modules/depots/depots.service.js [hunk 1]
```diff
@@ -61,6 +61,34 @@ function DepotService(Api, Modal, HttpCache) {
...
+   * @function getStockOutsForDate
+   *
+   * @description
+   * Looks for stock outs for a given depot on a given date.  If no date is provided, the
+   * current date is used.  This API is more efficient than the Stock.inventories.* API as
+   * it uses the stock_movement_status table.
+   */
+  service.getStockOutsForDate = function getStockOutsForDate(depotUuid, date = new Date()) {
+    const target = `/depots/${depotUuid}/flags/stock_out`;
+    return service.$http.get(target, { params : { date } })
```
Comment (mbayopanda) [inline client/src/modules/depots/depots.service.js:88]: 👍🏽 Excellent

File: client/src/modules/templates/bhStockPanelOutOfStock.tmpl.html [hunk 2]
```diff
@@ -1,12 +1,13 @@
 <div bh-stock-panel-out-of-stock ng-model-options="{ updateOn: 'default' }">
   <div class="panel panel-default">
     <div class="panel-heading">
-      <b><i class="{{ $ctrl.display }}"></i> <span translate> {{$ctrl.label}} </span></b>:
+      <b><i class="fa fa-battery-empty icon-out-of-stock"></i> <span translate> {{$ctrl.label}} </span></b>:
```
Comment (mbayopanda) [inline client/src/modules/templates/bhStockPanelOutOfStock.tmpl.html:4]: 👍🏽

File: client/src/modules/depots/depots.service.js [hunk 3]
```diff
@@ -61,6 +61,34 @@ function DepotService(Api, Modal, HttpCache) {
...
 
+  /**
+   * @function getStockQuantityForDate
+   *
+   * @description
+   * Looks up the quantites in stock for all inventory items for a depot.  This is more
+   * efficient than the more general Stock.inventories.* API since it uses the stock_movement_status
+   * table.
+   */
+  service.getStockQuantityForDate = function getStockQuantityForDate(depotUuid, date = new Date()) {
```
Comment (mbayopanda) [inline client/src/modules/depots/depots.service.js:72]: 👍🏽

**Review by mbayopanda — APPROVED** [review]
LGTM 👍🏽 

For the new API endpoints `/depots/:depotUuid/stock` and `/depots/:depotUuid/flags/stock_out` there is no Integration tests yet for testing data got from `stock_movement_status`

**Comment by jniles** [comment]
@mbayopanda good suggestion.  I'll implement tests for these routes.

**Comment by jniles** [comment]
@mbayopanda changes landed.

File: test/integration-stock/depots.js [hunk 4]
```diff
@@ -157,6 +153,27 @@ describe('(/depots) The depots API ', () => {
       .catch(helpers.handler);
   });
 
+  // mock today's date
+  const [today] = new Date().toISOString().split('T');
+
+  it('GET /depots/:uuid/stock returns the stock in a depot', async () => {
```
Comment (mbayopanda) [inline test/integration-stock/depots.js:159]: 👍🏽

File: test/integration-stock/depots.js [hunk 5]
```diff
@@ -157,6 +153,27 @@ describe('(/depots) The depots API ', () => {
...
+    helpers.api.listed(res, 3);
+
+    // check the quantites of each individual article
+    const [ampicilline, oxytocine, quinine] = res.body;
+    expect(ampicilline.quantity).to.equal(0);
+    expect(oxytocine.quantity).to.equal(9110);
+    expect(quinine.quantity).to.equal(360);
+  });
+
+  it('GET /depots/:uuid/flags/stock_out returns the stock in a depot', async () => {
```
Comment (mbayopanda) [inline test/integration-stock/depots.js:170]: 👍🏽

**Review by mbayopanda — APPROVED** [review]
It works fine, i have noticed some little thing about `Stock Out` in the search modal, we have two places for setting `Stock out` : 

image

image

In the first picture, when we define the stock out flag we do not have in the registry inventories which are in stock out status. The second picture gives us these inventories

image

My concerns is about, do we still need to have both place to set the `stock out` flag, we can just keep it in one place

**Comment by jniles** [comment]
What do you suggest?  How can we improve it?

**Comment by mbayopanda** [comment]
Actually the flag `Stock out` in the `Requete de recherche` tab seems broken while it is the best place according to me to set this flag.

The flag in the `Filtre par default` tab is more related with lots but it works perfectly

We can create a new PR that fix the `Stock out` in the `Requete de recherche` tab, and remove the other one

**Comment by jniles** [comment]
@mbayopanda Can you open an issue for the bug you found in `Requete de recherche`?  We should fix that.

If there isn't anything else to do in this PR, can we merge it?

**Pull request merged**